app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenProviderImpl.kt

```diff
@@ -48,56 +49,52 @@ object PoTokenProviderImpl : PoTokenProvider {
-            synchronized(WebPoTokenGenLock) {
-                val shouldRecreate = webPoTokenGenerator == null || forceRecreate ||
-                    webPoTokenGenerator!!.isExpired()
+            webPoTokenGenLock.withLock {
+                val gen = webPoTokenGenerator
+                val shouldRecreate = forceRecreate || gen == null || gen.isExpired
 
                 if (shouldRecreate) {
-
-                    val innertubeClientRequestInfo = InnertubeClientRequestInfo.ofWebClient()
-                    innertubeClientRequestInfo.clientInfo.clientVersion =
-                        YoutubeParsingHelper.getClientVersion()
-
-                    webPoTokenVisitorData = YoutubeParsingHelper.getVisitorDataFromInnertube(
-                        innertubeClientRequestInfo,
-                        NewPipe.getPreferredLocalization(),
-                        NewPipe.getPreferredContentCountry(),
-                        YoutubeParsingHelper.getYouTubeHeaders(),
-                        YoutubeParsingHelper.YOUTUBEI_V1_URL,
-                        null,
-                        false
-                    )
-                    // close the current webPoTokenGenerator on the main thread
-                    webPoTokenGenerator?.let { Handler(Looper.getMainLooper()).post { it.close() } }
+                    webPoTokenVisitorData = withContext(Dispatchers.IO) {
+                        val innertubeClientRequestInfo = InnertubeClientRequestInfo.ofWebClient()
+                        innertubeClientRequestInfo.clientInfo.clientVersion =
+                            YoutubeParsingHelper.getClientVersion()
+
+                        YoutubeParsingHelper.getVisitorDataFromInnertube(
+                            innertubeClientRequestInfo,
+                            NewPipe.getPreferredLocalization(),
+                            NewPipe.getPreferredContentCountry(),
+                            YoutubeParsingHelper.getYouTubeHeaders(),
+                            YoutubeParsingHelper.YOUTUBEI_V1_URL,
+                            null,
+                            false
+                        )
+                    }
+
+                    withContext(Dispatchers.Main) {
+                        webPoTokenGenerator?.close()
+                    }
 
                     // create a new webPoTokenGenerator
-                    webPoTokenGenerator = PoTokenWebView
-                        .newPoTokenGenerator(App.instance).blockingGet()
+                    webPoTokenGenerator = PoTokenWebView.getNewPoTokenGenerator(App.instance)
 
                     // The streaming poToken needs to be generated exactly once before generating
                     // any other (player) tokens.
-                    webPoTokenStreamingPot = webPoTokenGenerator!!
-                        .generatePoToken(webPoTokenVisitorData!!).blockingGet()
+                    webPoTokenStreamingPot = webPoTokenGenerator!!.generatePoToken(webPoTokenVisitorData!!)
```

Comment:
A bit weird that `webPaTokenGenerator` can be null here? Shouldn’t it be guaranteed to be there?